src/cart/crt_ctl.c

```diff
@@ -75,6 +75,135 @@ static int verify_ctl_in_args(struct crt_ctl_ep_ls_in *in_args)
 	return rc;
 }
 
+int
+crt_ctl_fill_buffer_cb(d_list_t *rlink, void *arg)
+{
+	struct crt_uri_item	*ui;
+	int			*idx;
+	char			*buf;
+	struct crt_uri_cache	*uri_cache = arg;
+	int			 i;
+	int			 uri_size;
+	int			 rc = 0;
+
+	D_ASSERT(rlink != NULL);
+	D_ASSERT(arg != NULL);
+
+	ui = crt_ui_link2ptr(rlink);
+
+	D_MUTEX_LOCK(&ui->ui_mutex);
+	for (i = 0; i < CRT_SRV_CONTEXT_NUM; i++) {
+		if (ui->ui_uri[i] != NULL) {
+			char *tmp_uri;
+
+			tmp_uri = ui->ui_uri[i];
+
+			idx = &uri_cache->info_idx;
+
+			buf = uri_cache->info_buf;
+
+			/* Pack rank */
+			*((d_rank_t *)(buf + *idx)) = ui->ui_rank;
```

Comment:
This code, through to L120 looks like it's packing a struct into a buffer, in which case why is it not using an array, and just doing assignment?  This seems needlessly complicated and error prone.